substrate/runtime/council/src/motions.rs

```diff
@@ -0,0 +1,350 @@
+
+/// Event for this module.
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize, Debug))]
+#[derive(Encode, Decode, PartialEq, Eq, Clone)]
+pub enum RawEvent<Hash, AccountId> {
+	/// A motion (given hash) has been proposed (by given account) with a threshold (given u32).
+	Proposed(AccountId, Hash, u32),
+	/// A motion (given hash) has been voted on by given account, leaving
+	/// a tally (yes votes and no votes given as u32s respectively).
+	Voted(AccountId, Hash, bool, u32, u32),
+	/// A motion was approved by the required threshold.
+	Approved(Hash),
+	/// A motion was not approved by the required threshold.
+	Disapproved(Hash),
+	/// A motion was executed; `bool` is true if returned without error.
+	Executed(Hash, bool),
+}
+
+impl<H, A> From<RawEvent<H, A>> for () {
+	fn from(_: RawEvent<H, A>) -> () { () }
+}
+
+decl_module! {
+	#[cfg_attr(feature = "std", serde(bound(deserialize = "<T as Trait>::Proposal: ::serde::de::DeserializeOwned")))]
+	pub struct Module<T: Trait> for enum Call where origin: <T as system::Trait>::Origin {
+		fn propose(origin, threshold: u32, proposal: Box<<T as Trait>::Proposal>) -> Result;
+		fn vote(origin, proposal: T::Hash, approve: bool) -> Result;
+	}
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as CouncilMotions {
+		/// The (hashes of) the active proposals.
+		pub Proposals get(proposals): default Vec<T::Hash>;
+		/// Actual proposal for a given hash, if it's current.
+		pub ProposalOf get(proposal_of): map [ T::Hash => <T as Trait>::Proposal ];
+		/// Votes for a given proposal: (required_yes_votes, yes_voters, no_voters).
+		pub Voting get(voting): required map [ T::Hash => (u32, Vec<T::AccountId>, Vec<T::AccountId>) ];
+	}
+}
+
+impl<T: Trait> Module<T> {
+
+	/// Deposit one of this module's events.
+	fn deposit_event(event: Event<T>) {
+		<system::Module<T>>::deposit_event(<T as Trait>::Event::from(event).into());
+	}
+
+	pub fn is_councillor(who: &T::AccountId) -> bool {
+		<Council<T>>::active_council().iter()
+			.any(|&(ref a, _)| a == who)
+	}
+
+	// Dispatch
+	fn propose(origin: <T as system::Trait>::Origin, threshold: u32, proposal: Box<<T as Trait>::Proposal>) -> Result {
+		let who = ensure_signed(origin)?;
+
+		ensure!(Self::is_councillor(&who), "proposer not on council");
+
+		let proposal_hash = T::Hashing::hash_of(&proposal);
+
+		ensure!(!<ProposalOf<T>>::exists(proposal_hash), "duplicate proposals not allowed");
+
+		<Proposals<T>>::mutate(|proposals| proposals.push(proposal_hash));
+		<ProposalOf<T>>::insert(proposal_hash, *proposal);
+		<Voting<T>>::insert(proposal_hash, (threshold, vec![who.clone()], vec![]));
+
+		Self::deposit_event(RawEvent::Proposed(who, proposal_hash, threshold));
+
+		Ok(())
+	}
+
+	fn vote(origin: <T as system::Trait>::Origin, proposal: T::Hash, approve: bool) -> Result {
+		let who = ensure_signed(origin)?;
+
+		ensure!(Self::is_councillor(&who), "voter not on council");
+
+		let mut voting = Self::voting(&proposal);
```

Comment:
Consider there are two extrinsics that vote for the same proposal. The first one will lead to execution of proposal. What would happen with the second one?

Reply:
panic, and indeed, that should just be a bail.